features/mbedtls/targets/TARGET_STM/ccm_alt.c

```diff
@@ -0,0 +1,511 @@
+    cryp_context_count++;
+    __enable_irq();
+
+    cryp_zeroize((void *)ctx, sizeof(mbedtls_ccm_context));
+
+#if (MBEDTLS_DEBUG)
+    printf("[ALT] mbedtls_ccm_init %u\n", cryp_context_count);
+#endif
+}
+
+int mbedtls_ccm_setkey(mbedtls_ccm_context *ctx,
+                       mbedtls_cipher_id_t cipher,
+                       const unsigned char *key,
+                       unsigned int keybits)
+{
+    unsigned int i;
+    int ret = 0;
+
+#if (MBEDTLS_DEBUG)
+    printf("[ALT] mbedtls_ccm_setkey\n");
+#endif
+
+    CCM_VALIDATE_RET(ctx != NULL);
+    CCM_VALIDATE_RET(key != NULL);
+
+    /* Protect context access                                  */
+    /* (it may occur at a same time in a threaded environment) */
+#if defined(MBEDTLS_THREADING_C)
+    if (mbedtls_mutex_lock(&cryp_mutex) != 0) {
+        return (MBEDTLS_ERR_THREADING_MUTEX_ERROR);
+    }
+#endif /* MBEDTLS_THREADING_C */
+
+    switch (keybits) {
+        case 128:
+            ctx->hcryp_ccm.Init.KeySize = CRYP_KEYSIZE_128B;;
+            break;
+
+        case 192:
+#if ( USE_AES_KEY192 == 1 )
+            ctx->hcryp_ccm.Init.KeySize = CRYP_KEYSIZE_192B;
+            break;
+#else
+            ret = MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED;
+            goto exit;
+#endif /* USE_AES_KEY192 */
+
+        case 256:
+            ctx->hcryp_ccm.Init.KeySize = CRYP_KEYSIZE_256B;
+            break;
+
+        default :
+            ret = MBEDTLS_ERR_CCM_BAD_INPUT;
+            goto exit;
+    }
+
+    /* Format and fill AES key  */
+    for (i = 0; i < (keybits / 32) ; i++) {
+        GET_UINT32_BE(ctx->ccm_key[i], key, 4 * i);
+    }
+
+    /* include the appropriate instance name */
+#if defined (AES)
+    ctx->hcryp_ccm.Instance = AES;
+#elif defined (AES1)
+    ctx->hcryp_ccm.Instance = AES1;
+#else /* CRYP */
+    ctx->hcryp_ccm.Instance = CRYP;
+#endif /* AES */
+
+    ctx->hcryp_ccm.Init.DataType = CRYP_DATATYPE_8B;
+    ctx->hcryp_ccm.Init.pKey = ctx->ccm_key;
+    ctx->hcryp_ccm.Init.pInitVect  = NULL;
+    ctx->hcryp_ccm.Init.Algorithm  = CRYP_AES_CCM;
+    ctx->hcryp_ccm.Init.Header     = NULL;
+    ctx->hcryp_ccm.Init.HeaderSize = 0;
+    ctx->hcryp_ccm.Init.B0 = NULL;
+    ctx->hcryp_ccm.Init.DataWidthUnit = CRYP_DATAWIDTHUNIT_BYTE;
+
+    if (HAL_CRYP_Init(&ctx->hcryp_ccm) != HAL_OK) {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+        goto exit;
+    }
+
+    /* allow multi-context of CRYP : save context */
+    ctx->ctx_save_cr = ctx->hcryp_ccm.Instance->CR;
+
+exit :
+    /* Free context access */
+#if defined(MBEDTLS_THREADING_C)
+    if (mbedtls_mutex_unlock(&cryp_mutex) != 0) {
+        ret = MBEDTLS_ERR_THREADING_MUTEX_ERROR;
+    }
+#endif /* MBEDTLS_THREADING_C */
+
+    return (ret);
+}
+
+/*
+ * Free context
+ */
+void mbedtls_ccm_free(mbedtls_ccm_context *ctx)
+{
+#if (MBEDTLS_DEBUG)
+    printf("[ALT] mbedtls_ccm_free %u\n", cryp_context_count);
+#endif
+
+    if (ctx == NULL) {
+        return;
+    }
+
+    __disable_irq();
```

Comment:
Any weaker synchronization we can use here?

Here any any other places we use `__disable_irq()`.